### ISSUE TITLE
There are lots of make issue when I build the kernel

### ISSUE THREAD

**Issue by Gumi-presentation-by-Dzh** [nehaag/thermostat_asplos_2017#2]
like

24 In file included from mm/swap.c:36:0:
 25 include/linux/page_idle.h:87:20: error: redefinition of ‘set_page_young’
 26  static inline void set_page_young(struct page *page)
 27                     ^
 28 In file included from include/linux/page_ref.h:6:0,
 29                  from include/linux/mm.h:25,
 30                  from mm/swap.c:16:
 31 include/linux/page-flags.h:405:20: note: previous definition of ‘set_page_young’     was here
 32  static inline void set_page_young(struct page *page) {}
 33                     ^
 34 In file included from mm/swap.c:36:0:
 35 include/linux/page_idle.h:107:20: error: redefinition of ‘clear_page_idle’
 36  static inline void clear_page_idle(struct page *page)
 37                     ^
 38 In file included from include/linux/page_ref.h:6:0,
 39                  from include/linux/mm.h:25,
 40                  from mm/swap.c:16:
 41 include/linux/page-flags.h:406:20: note: previous definition of ‘clear_page_idle    ’ was here
 42  static inline void clear_page_idle(struct page *page) {}
 43                     ^
 44 make[1]: *** [mm/swap.o] Error 1
 45 make[1]: *** Waiting for unfinished jobs....
 46 make: *** [mm] Error 2
 47 make: *** Waiting for unfinished jobs....
 48 In file included from fs/proc/task_mmu.c:16:0:
 49 include/linux/page_idle.h:87:20: error: redefinition of ‘set_page_young’
 50  static inline void set_page_young(struct page *page)
 51                     ^
 52 In file included from include/linux/page_ref.h:6:0,
 53                  from include/linux/mm.h:25,
 54                  from fs/proc/task_mmu.c:1:

So, what's config do you use? Can this kernel work to test?

**Comment by Gumi-presentation-by-Dzh** [comment]
When I set the config file to let the bzImage built and then I reboot, there is a Kernel panic-not syncing fatal exception happened in my server starting. It happend after "uncore_event_init_cpu & intel_uncore_int &uncore_cpu_setup"…………until “rest_init & uncore_change_context”.

SO what wrong with this kernel configuration?

**Comment by Dwade3** [comment]
Hi,I had the same problem too. Have you solved it ?

**Comment by Gumi-presentation-by-Dzh** [comment]
Do you mean the make error or fatal exception?

**Comment by Dwade3** [comment]
make error

**Comment by Gumi-presentation-by-Dzh** [comment]
You can set the config file to solve this problem. But the fatal exception in kernel is still a problem for me.

**Comment by Gumi-presentation-by-Dzh** [comment]
First you need to add CONFIG_KSTALED to you .config file as I use the kernel 3.12 config file does not have this configure.

**Comment by Gumi-presentation-by-Dzh** [comment]
The other config you need add maybe CONFIG_POISON_PAGE as I remember. In this way you will have the right bzImage for kernel model. If you success, let me know, thanks.

**Comment by Dwade3** [comment]
Do you mean  i should add CONFIG_POISON_PAGE and CONFIG_KSTALED to my .config file manually?

**Comment by Gumi-presentation-by-Dzh** [comment]
Yes

**Comment by Gumi-presentation-by-Dzh** [comment]
At least I succeeded in compiling bzimage by adding these two settings in .config

**Comment by Gumi-presentation-by-Dzh** [comment]
If you success this kernel, let me know, thanks.

**Comment by Dwade3** [comment]
No problem. I will try. Thanks for your time.

**Comment by Dwade3** [comment]
Excuse me. Could you tell me what make command you exactly use? I add CONFIG_KSTALED to my .config file, but it will disappear when i make.

**Comment by Dwade3** [comment]
Excuse me. I firstly add CONFIG_KSTALED to my .config file,but it will disappear when i execute make command.Then, i add config KSTALED in Kconfig file.And there are new errors as below:
arch/x86/mm/fault.c:1198:32: error: implicit declaration of function ‘get_mem_cgroup_from_mm’ [-Werror=implicit-function-declaration]
     struct mem_cgroup *memcg = get_mem_cgroup_from_mm(current->mm);
                                ^
arch/x86/mm/fault.c:1198:32: warning: initialization makes pointer from integer without a cast [-Wint-conversion]
arch/x86/mm/fault.c:1199:19: error: dereferencing pointer to incomplete type ‘struct mem_cgroup’
     css_put(&memcg->css);
                   ^
arch/x86/mm/fault.c:1200:37: error: implicit declaration of function ‘mem_cgroup_poison_page’ [-Werror=implicit-function-declaration]
     int is_mem_cgroup_poison_page = mem_cgroup_poison_page(memcg);
                                     ^
arch/x86/mm/fault.c:1200:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
     int is_mem_cgroup_poison_page = mem_cgroup_poison_page(memcg);
     ^
arch/x86/mm/fault.c:1409:9: error: implicit declaration of function ‘mem_cgroup_increment_fault_time’ [-Werror=implicit-function-declaration]
         mem_cgroup_increment_fault_time(memcg, end_time-start_time);
         ^
arch/x86/mm/fault.c: At top level:
arch/x86/mm/fault.c:694:1: warning: ‘pgtable_bad’ defined but not used [-Wunused-function]
 pgtable_bad(struct pt_regs *regs, unsigned long error_code

Have you ever met these?

**Comment by Gumi-presentation-by-Dzh** [comment]
Did you add CONFIG_POISON_PAGE?

**Comment by Dwade3** [comment]
Yes, all added to Kconfig

**Comment by Gumi-presentation-by-Dzh** [comment]
We didn't meet this error , we use the config file in /boot and use cp command to "cp config .config" to build the .config file and then use the  sh -c 'yes "" | make oldconfig' to set the oldconfig.

**Comment by Dwade3** [comment]
sudo make
scripts/kconfig/conf  --silentoldconfig Kconfig
  CHK     include/config/kernel.release
  CHK     include/generated/uapi/linux/version.h
  CHK     include/generated/utsrelease.h
  CC      kernel/bounds.s
In file included from kernel/bounds.c:9:0:
include/linux/page-flags.h:121:5: error: redeclaration of enumerator ‘PG_young’
     PG_young,  /* kstaled cleared pte_young */
     ^~~~~~~~
include/linux/page-flags.h:117:2: note: previous definition of ‘PG_young’ was here
  PG_young,
  ^~~~~~~~
include/linux/page-flags.h:122:5: error: redeclaration of enumerator ‘PG_idle’
     PG_idle,  /* idle since start of kstaled interval */
     ^~~~~~~
include/linux/page-flags.h:118:2: note: previous definition of ‘PG_idle’ was here
  PG_idle,
  ^~~~~~~
I did as you say. But make still doesn't work. What make command do you use?

**Comment by Gumi-presentation-by-Dzh** [comment]
sudo make bzImage

As you mentioned , redeclaration of enumerator ‘PG_young’ (/* kstaled cleared pte_young /) I think maybe still the config error.

You can go to page-flags.h to see the PG_young's config define to set the configure in .config fie.

I just set these to make sure there is not redeclaration in make process.
CONFIG_KSTALED=y  #to bring define
CONFIG_POISON_PAGE=y   #to bring define

Hope this will help.

**Comment by Dwade3** [comment]
I succeed to compile this kernel and install it on Ubuntu16.04. But i still don't know how to analyze and use this kernel.

**Comment by Gumi-presentation-by-Dzh** [comment]
Do you mean it run succeed as well? What's your config?

**Comment by Dwade3** [comment]
Yeah, i execute make modules_install and install after compiling. Then i simply modify grub configuration. And it appears to successfully install 4.5.0 kernel after reboot.
But I change idle_page_tracking to no in config file when i compile , i don't know if this is right.

**Comment by Dwade3** [comment]
Do you know how to evaluate this kernel? Run those benchmarks in virtual machines as the corresponding paper says?

**Comment by Gumi-presentation-by-Dzh** [comment]
Sorry to keep you waiting, because I can not run this kernel successful so I can not give you any advice. In that paper, you should run your APP in KVM on tmfs. I will try your configure file in the future if i am free in time. So keep contact if there are more problems we find.
You mean that you use  Ubuntu and add CONFIG_KSTALED , CONFIG_POISON_PAGE and also change the idle_page_tracking to no to build the kernel when you compile it?

**Comment by Gumi-presentation-by-Dzh** [comment]
Maybe the CentOS7 does not support this kernel? We will try it in your configuration and find the way to test this work.

**Comment by Dwade3** [comment]
Yes. I see you mentioned that you used the config file of version 3.12. I think maybe you should compile this kernel on the OS that has the kernel of newer version.
Thanks for all your help. Let's keep in touch.

**Comment by Gumi-presentation-by-Dzh** [comment]
Thanks a lot.

**Comment by Gumi-presentation-by-Dzh** [comment]
The last question , you say that the config file we used maybe a little old , so what version of configuration file you use?

**Comment by Gumi-presentation-by-Dzh** [comment]
And did you change something of that configuration file you used?

**Comment by Dwade3** [comment]
linux kernel 4.4.44. Ubuntu16.04

**Comment by Dwade3** [comment]
I firstly added CONFIG_KSTALED , CONFIG_POISON_PAGE  to my config as you told me. But there are still errors while compiling. So I change idle_page_tracking to no. And it works.

**Comment by Gumi-presentation-by-Dzh** [comment]
You mean that the configuration file  you used is in /boot/ ?

**Comment by Dwade3** [comment]
Yes

**Comment by Gumi-presentation-by-Dzh** [comment]
gotcha

**Comment by Gumi-presentation-by-Dzh** [comment]
I will try to install the kernel with a high version of config, if there is a new problem and discovery, keep in touch。

**Comment by Gumi-presentation-by-Dzh** [comment]
@Dwade3 Do you test this successfully？

**Comment by Dwade3** [comment]
No, it just didn't work. Sorry to reply this late
